### PR TITLE
push to stable

### PR DESCRIPTION
**Pull request by vanhauser-thc**
maybe merge?
not much there I think but why wait :)

travis is still not green ... unknown why it fails the test

### CODE REVIEW

File: include/afl-fuzz.h [hunk 1]
```diff
@@ -1001,6 +1003,32 @@ static inline u32 rand_below(afl_state_t *afl, u32 limit) {
 
 }
 
+/* we prefer lower range values here */
+/* this is only called with normal havoc, not MOpt, to have an equalizer for
+   expand havoc mode */
+static inline u32 rand_below_datalen(afl_state_t *afl, u32 limit) {
```
Comment (vanhauser-thc) [inline include/afl-fuzz.h:1010]: currently unused (again) but I would like to keep it available

**Comment by vanhauser-thc** [comment]
why is travis green? how did that travis specific float cmpcov bug go away? weird.

I would like to merge as our code reviews are now quite some days back and I want to add improvements into dev.
Or is anyone working on something that should merge now?

**Review by domenukk — APPROVED** [review]

**Comment by vanhauser-thc** [comment]
can we merge this? especially @andreafioraldi

**Comment by andreafioraldi** [comment]
For the linking stuffs yes, I widely tested it. However I'm not sure that neverzero in pcguard will produce efficient code, did you checked? If not, we can copy the inline asm snippet of qemu mode, at least for x86.

**Comment by vanhauser-thc** [comment]
@andreafioraldi 

> For the linking stuffs yes, I widely tested it. However I'm not sure that neverzero in pcguard will produce efficient code, did you checked? If not, we can copy the inline asm snippet of qemu mode, at least for x86.

yes I checked it.

before:
```
Dump of assembler code for function __sanitizer_cov_trace_pc_guard:
   0x0000000000000ac0 <+0>:	mov    rax,QWORD PTR [rip+0x0]        # 0xac7 <__sanitizer_cov_trace_pc_guard+7>
   0x0000000000000ac7 <+7>:	mov    rax,QWORD PTR [rax]
   0x0000000000000aca <+10>:	mov    ecx,DWORD PTR [rdi]
   0x0000000000000acc <+12>:	add    BYTE PTR [rax+rcx*1],0x1
   0x0000000000000ad0 <+16>:	ret    
End of assembler dump.
```
after
```
gdb$ disass __sanitizer_cov_trace_pc_guard
Dump of assembler code for function __sanitizer_cov_trace_pc_guard:
   0x0000000000000b80 <+0>:	mov    rax,QWORD PTR [rip+0x0]        # 0xb87 <__sanitizer_cov_trace_pc_guard+7>
   0x0000000000000b87 <+7>:	mov    rax,QWORD PTR [rax]
   0x0000000000000b8a <+10>:	mov    ecx,DWORD PTR [rdi]
   0x0000000000000b8c <+12>:	mov    dl,BYTE PTR [rax+rcx*1]
   0x0000000000000b8f <+15>:	add    dl,0x1
   0x0000000000000b92 <+18>:	adc    dl,0x0
   0x0000000000000b95 <+21>:	mov    BYTE PTR [rax+rcx*1],dl
   0x0000000000000b98 <+24>:	ret    
End of assembler dump.
```
which looks fine IMHO.

fuzzbench showed that neverzero is "usually" very slightly better than wrapping counters.

what do you think?

**Comment by andreafioraldi** [comment]
If this is good also for arm64, ok for me.

**Comment by vanhauser-thc** [comment]
I will check it out in a few minutes

**Comment by vanhauser-thc** [comment]
This is for aarch64 - before:
```
Dump of assembler code for function __sanitizer_cov_trace_pc_guard:
   0x0000000000000aa8 <+0>:	adrp	x8, 0x0 <send_forkserver_error>
   0x0000000000000aac <+4>:	ldr	x8, [x8]
   0x0000000000000ab0 <+8>:	ldr	w9, [x0]
   0x0000000000000ab4 <+12>:	ldr	x8, [x8]
   0x0000000000000ab8 <+16>:	ldrb	w10, [x8, x9]
   0x0000000000000abc <+20>:	add	w10, w10, #0x1
   0x0000000000000ac0 <+24>:	strb	w10, [x8, x9]
   0x0000000000000ac4 <+28>:	ret
```
after:
```
Dump of assembler code for function __sanitizer_cov_trace_pc_guard:
   0x0000000000000ab0 <+0>:	adrp	x8, 0x0 <send_forkserver_error>
   0x0000000000000ab4 <+4>:	ldr	x8, [x8]
   0x0000000000000ab8 <+8>:	ldr	w9, [x0]
   0x0000000000000abc <+12>:	ldr	x8, [x8]
   0x0000000000000ac0 <+16>:	ldrb	w10, [x8, x9]
   0x0000000000000ac4 <+20>:	add	w10, w10, #0x1
   0x0000000000000ac8 <+24>:	add	w10, w10, w10, lsr #8
   0x0000000000000acc <+28>:	strb	w10, [x8, x9]
   0x0000000000000ad0 <+32>:	ret
```

looks even better IMHO

**Pull request merged**